snuba/pipeline/query_plan_delegator.py

```diff
@@ -27,13 +53,68 @@ def __init__(
         runner: QueryRunner,
         query_plan_builders: QueryPlanBuilders,
         selector_func: SelectorFunc,
-        callback_func: CallbackFunc,
+        callback_func: Optional[CallbackFunc],
     ):
         self.__request = request
+        self.__runner = runner
         self.__query_plan_builders = query_plan_builders
+        self.__selector_func = selector_func
+        self.__callback_func = callback_func
 
     def execute(self) -> QueryResult:
-        raise NotImplementedError
+        def execute_query(
+            query_plan_builder: ClickhouseQueryPlanBuilder,
+        ) -> QueryResult:
+
+            query_plan = query_plan_builder.build_plan(self.__request)
+
+            _execute_query_plan_processors(query_plan, self.__request)
+
+            return query_plan.execution_strategy.execute(
+                query_plan.query, self.__request.settings, self.__runner
+            )
```

Comment:
So I think we have a problem here for subqueries and joins.
subqueries and joins need to work this way:
1) we receive an ugly composite query
2) we split it into multiple subqueries 
3) each of that has to go through the processing pipeline (excluding the execution step)
4) when these pipelines are done a pipeline step needs to recompose the preocessed subqueries into a composite query
5) additional composite query optimization processing could happen
6) execute the query through a composite query strategy.

This means decoupling the execution step from the processing step.

I think there are two problem we need to find a viable solution here:
1) It seems possible that a pipeline will be individual till a point and then fork aftewards. Which means how do we make a composite query pipeline trigger two executions oof the composite query when it is an individual subquery processing that decides the pipeline has to fork? Should we allow pipelines to fork dynamically or can we know this statically when defining the pipeline? I need to put more thoughts on this, do you have ideas ?

2) Can we instead not support the two at the same time? Plan delegator will never be picked if the query is composite? In that case how does the pipeline builder look like ?

Reply:
In its current form QueryPlanDelegator doesn't support subqueries and joins (mostly since there is no immediate use case to use them together, and I didn't really know how joins will end up looking like). However this means that the delegator is much less useful than it could potentially be in future. Thinking about this again, I think it would be better and potentially also simpler to make the delegator a PipelineDelegator - a special type of pipeline that can be passed multiple other query pipelines which are executed in parallel. So since steps 1-6 that you mentioned above are all steps in the pipeline, they would get executed once for for each pipeline being "selected". That way we could use the PipelineDelegator to switch out and test any parts of the pipeline (not just the query plan and storage part like it does now).

What do you think?

Reply:
I like the idea of the pipeline delegator. Though is there a way to drive the storage selection (which is what makes the two pipelines different) at the beginning of the pipeline ?

Reply:
What do you mean by "at the beginning"? Since the storage selector would be unique for each pipeline passed to the delegator, it would be set when the query plan for that pipeline is built.  A pipeline either has one storage (if using `SingleStorageQueryPlanBuilder`) or multiple (if using `SelectedStorageQueryPlanBuilder` which is the case for the storages that have readonly variants)

Reply:
By "at the beginning" I mean that the split of a composite query into multiple subqueries that gets processed independently and then packaged back happens before we run entity processor. While your execution pipeline starts at the query plan building, which is after the entity query processing.
For a composite query (expanding a bit on the steps I mentioned above):
- we split it into multiple subqueries
- here each subquery is processed independently (each subquery is a simple query on an entity. It includes: entity validators, entity query processors, storage query plan building, storage processor execution,)
- then we package these processed query back into a composite clickhouse query
- then we execute it.

if we managed to make your pipeline start before the composite query split, then we have no problem. But, let's say we did that, how do we convince the query plan builder to pick onne storage instead of another when we want to run one pipeline on errors and one on events, if the pipeline starts in the entity processing phase (before we pick a storage) ?
Do you have an idea about this that allows us to do that without having to fork the main pipeline (the one that runs the composite query) due to a decision made by onne of the sub-pipelines (the events one) ?

Reply:
It's my understanding that supporting joins/composite queries via splitting and recombining subqueries was specifically the main reason for creating the pipeline abstraction to begin with. So this PR is based on the idea that all of the steps mentioned are completely contained within each pipeline that is passed to the delegator.

Are you suggesting as an alternative placing splitting outside of the pipeline, but processing and merging back the previously split queries and execution inside the pipeline? (i.e. a pipeline receives as input multiple split queries and outputs the query result)

Reply:
Discussed offline.
The  individual pipeline will be extended (whether in this PR or not) to cover the full query processing.
The delegator will run two complete individual pipelines, these will be customized by providing them a way to build the query plan that is decided by the delegator.
at that point this will be extensible to composite queries as long as the delegator will be able to propagate the decision on the query plan to build down to the individual pipeline.